bench/utils/__init__.py

```diff
@@ -53,32 +53,33 @@ def is_frappe_app(directory: str) -> bool:
 def is_valid_frappe_branch(frappe_path:str, frappe_branch:str):
 	""" Check if a branch exists in a repo. Throws InvalidRemoteException if branch is not found
 
-	Uses github's api without auth to query branch.
-	If rate limited by gitapi, requests are sent to github.com
+	Uses native git command to check for branches on a remote.
 
 	:param frappe_path: git url
 	:type frappe_path: str
 	:param frappe_branch: branch to check
 	:type frappe_branch: str
 	:raises InvalidRemoteException: branch for this repo doesn't exist
 	"""
-	if "http" in frappe_path and frappe_branch:
-		frappe_path = frappe_path.replace(".git", "")
+	import subprocess
+
+	message = f"Invalid branch {frappe_branch} for remote {frappe_path}"
+	if frappe_branch:
 		try:
-			owner, repo = frappe_path.split("/")[3], frappe_path.split("/")[4]
-		except IndexError:
-			raise InvalidRemoteException("Invalid git url")
-		git_api_req = f"https://api.github.com/repos/{owner}/{repo}/branches"
-		res = requests.get(git_api_req).json()
-
-		if "message" in res:
-			# slower alternative with no rate limit
-			github_req = f'https://github.com/{owner}/{repo}/tree/{frappe_branch}'
-			if requests.get(github_req).status_code != 200:
-				raise InvalidRemoteException("Invalid git url")
-
-		elif frappe_branch not in [x["name"] for x in res]:
-			raise InvalidRemoteException("Frappe branch does not exist")
+			ret = subprocess.check_output(
+				(
+					"git",
+					"ls-remote",
+					"--heads",
+					frappe_path,
+					frappe_branch,
+				),
+				encoding="UTF-8",
+			)
+			if not ret:
+				raise InvalidRemoteException(message)
+		except subprocess.CalledProcessError:
+			raise InvalidRemoteException(message)
```

Comment:
the message here should be different?

We don't get any output when the git url is correct but branch is incorrect/does not exist,
though we do get an error when the git url is wrong - so the message over here should reflect the same? 
maybe something like: "The provided frappe path/git url is invalid"

![Screenshot 2022-04-23 at 5 57 06 PM](https://user-images.githubusercontent.com/32034600/164894447-be29d104-e766-4eef-b550-7586f438ae2b.png)

Reply:
implemented by [bb10914](https://github.com/frappe/bench/pull/1294/commits/bb10914c435b86d65c1ee8080a70d9fcb3af7f3a)